assets/app/scripts/controllers/create.js

```diff
@@ -47,33 +119,198 @@ angular.module('openshiftConsole')
       return false;
     }
 
-    function updateTemplates() {
-      // Check if we've loaded templates from both the openshift and project namespaces.
-      $scope.templatesLoaded = openshiftTemplates && projectTemplates;
-      $scope.instantApps = {};
-      $scope.otherTemplates = {};
-
-      // Categorize templates as instant apps or "other."
-      var categorizeTemplates = function(template) {
-        var uid = uidFilter(template);
-        if (isInstantApp(template)) {
-          $scope.instantApps[uid] = template;
+    function matchesKeyword(name, tags, keyword) {
+      // Match tag substrings when comparing keywords.
+      if (hasTag(keyword, tags, true)) {
+        return true;
+      }
+
+      return name.toLowerCase().indexOf(keyword) !== -1;
+    }
+
+    // Return true only if item matches both tag and keyword filters when both are set.
+    function matchesFilter(name, tags) {
+      var keywords, i;
+
+      if ($scope.filter.tag && !hasTag($scope.filter.tag, tags)) {
+        return false;
+      }
+
+      if ($scope.filter.keyword) {
+        keywords = $scope.filter.keyword.split(/\s+/);
+        for (i = 0; i < keywords.length; i++) {
+          if (!matchesKeyword(name, tags, keywords[i])) {
+            return false;
+          }
+        }
+      }
+
+      return true;
+    }
+
+    function countItems(category) {
+      var builders = $scope.filteredBuildersByCategory[category] || [],
+        templates = $scope.filteredTemplatesByCategory[category] || [];
+
+      return Math.max(builders.length, $scope.itemLimit) + Math.max(templates.length, $scope.itemLimit);
+    }
+
+    // Keep the more important categories at the top of each column, but try to
+    // keep roughly the same number of visible items in each column.
+    function updateColumns() {
+      var i,
+          numLeft = 0,
+          numRight = 0,
+          categories = $scope.filteredCategoryTags;
+
+      $scope.leftCategories  = [];
+      $scope.rightCategories = [];
+
+      for (i = 0; i < categories.length; i++) {
```

Comment:
either angular.forEach or var category = categories[i] so you dont have to repeat categories[i]